public/js/controllers/chat.controller.js

```diff
@@ -51,36 +54,114 @@
                 vm.loggedIn = true;
                 vm.user = userResult.data;
                 chatDataService.getUsers().then(function(result) {
-                    vm.users = result.data;
-                });
-                setInterval(function() {
-                    vm.conversations.forEach(function(conversation) {
-                        refreshMessages(conversation.data.id);
+                    result.data.forEach(function(user) {
+                        displayUser(user);
                     });
-                }, 333);
+                }).then(function() {
+                    pollNotifications();
+                    $interval(pollNotifications, 500);
+                });
             }).catch(function() {
                 chatDataService.getOAuthUri().then(function(result) {
                     vm.loginUri = result.data.uri;
                 });
             });
         }
 
+        // Set of functions for handling different notification types
+        var notificationHandler = {
+            "new_messages": function(notificationData) {
+                if (vm.conversations[notificationData.conversationID]) {
+                    // If we have the conversation open, refresh it
+                    return refreshConversation(notificationData.conversationID);
+                } else {
+                    // Otherwise set the unread message count for the user
+                    return chatDataService.getConversationMessages(notificationData.conversationID, {
+                        countOnly: true,
+                        timestamp: notificationData.since
+                    }).then(function(countResponse) {
+                        var userIdx = vm.users.findIndex(function(currentUser) {
+                            return currentUser.data.id === notificationData.otherID;
+                        });
+                        if (userIdx !== -1) {
+                            vm.users[userIdx].unreadMessageCount = countResponse.data.count;
+                        }
+                    });
+                }
+            }
+        };
+
+        // Performs an action based on the received notification; this action does not necessarily "resolve" the
+        // notification, but should be idempotent if the server has not changed state since the notification was
```

Comment:
It's wrong to call this function idempotent, so I would leave that out. Just say that no changes are made if the server state hasn't updated since the last notification, or similar.